src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_code.go

```diff
@@ -172,62 +175,56 @@ func (a *StoragePowerActorCode_I) GetTotalPower(rt Runtime) block.StoragePower {
 	return totalPower
 }
 
-func (a *StoragePowerActorCode_I) EnsurePledgeCollateralSatisfied(rt Runtime) bool {
-
-	ret := false
-
-	// TODO: convert msgSender to MinerActorID
-	var minerID addr.Address
+func (a *StoragePowerActorCode_I) EnsurePledgeCollateralSatisfied(rt Runtime) InvocOutput {
 
 	h, st := a.State(rt)
+	ret := st._ensurePledgeCollateralSatisfied(rt)
+	UpdateRelease(rt, h, st)
 
-	powerEntry, found := st.PowerTable()[minerID]
-
-	if !found {
-		rt.Abort("miner not found.")
-	}
-
-	pledgeCollateralRequired := st._getPledgeCollateralReq(rt, powerEntry.ActivePower()+powerEntry.InactivePower())
-
-	if pledgeCollateralRequired < powerEntry.LockedPledgeCollateral() {
-		ret = true
-	} else if pledgeCollateralRequired < (powerEntry.LockedPledgeCollateral() + powerEntry.AvailableBalance()) {
-		st._lockPledgeCollateral(rt, minerID, (pledgeCollateralRequired - powerEntry.LockedPledgeCollateral()))
-		ret = true
+	if !ret {
+		return rt.ErrorReturn(exitcode.InsufficientPledgeCollateral)
 	}
 
-	UpdateRelease(rt, h, st)
-
-	return ret
+	return rt.SuccessReturn()
 }
 
-func (a *StoragePowerActorCode_I) ProcessFaultReport(rt Runtime, report sector.FaultReport) {
+// slash pledge collateral for Declared, Detected and Terminated faults
+func (a *StoragePowerActorCode_I) SlashPledgeForStorageFault(rt Runtime, affectedPower block.StoragePower, faultType sector.StorageFaultType) {
```

Comment:
In all of these methods, we should validate the caller (regardless of whether we need to interact with the caller); this is necessary for correctness. (`ValidateImmediateCallerIs`, `ValidateImmediateCallerMatches`, `ValidateImmediateCallerAcceptAny`)

Reply:
@jzimmerman how do we validate that a caller is a StorageMinerAddr since it is a non-singleton actor?

Reply:
`ActorState::CodeID` and the associated methods in `actor.go` should handle it, I think. (There might not be a clean way to get the ActorState of the sender yet, though -- let me know if you hit this issue and I can add some functionality to the runtime API.)